# 1. Tratamento dos dados do ENEM 2019 do estado de São Paulo

In [1]:
import pandas as pd
import numpy as np

In [4]:
dados = pd.read_csv(r"C:\Users\054150\OneDrive - Construtora Barbosa Mello SA\microdados_enem_2019_sp.csv", sep=';', encoding='iso-8859-1')

# 2. Limpeza dos dados

In [6]:
dados.head()

NU_INSCRICAO  CO_MUNICIPIO_RESIDENCIA NO_MUNICIPIO_RESIDENCIA  \
0     1,90E+11                  3546801            Santa Isabel   
1     1,90E+11                  3538204             Pinhalzinho   
2     1,90E+11                  3550308               São Paulo   
3     1,90E+11                  3548708   São Bernardo do Campo   
4     1,90E+11                  3549409    São Joaquim da Barra   

   CO_UF_RESIDENCIA SG_UF_RESIDENCIA  NU_IDADE TP_SEXO  TP_ESTADO_CIVIL  \
0                35               SP        22       M                1   
1                35               SP        19       M                1   
2                35               SP        17       F                1   
3                35               SP        19       M                1   
4                35               SP        37       M                2   

   TP_COR_RACA  TP_NACIONALIDADE  ...   NU_NOTA_MT TP_LINGUA  \
0            3                 1  ...       6002.0         0   
1            3                 1  ...          NaN         1   
2            3                 1  ...       7318.0         0   
3            3                 1  ...       3932.0         1   
4            1                 1  ...          NaN         1   

   TP_STATUS_REDACAO NU_NOTA_COMP1  NU_NOTA_COMP2  NU_NOTA_COMP3  \
0                1.0         160.0          200.0          180.0   
1                NaN           NaN            NaN            NaN   
2                1.0         160.0          120.0          200.0   
3                1.0          80.0          160.0          120.0   
4                NaN           NaN            NaN            NaN   

   NU_NOTA_COMP4  NU_NOTA_COMP5  NU_NOTA_REDACAO  Q025  
0          200.0          200.0            940.0     B  
1            NaN            NaN              NaN     A  
2          200.0          200.0            880.0     A  
3          100.0          100.0            560.0     B  
4            NaN            NaN              NaN     B  

[5 rows x 44 columns]

In [7]:
# numero total de inscritos em são paulo

dados.shape

(813772, 44)

In [9]:
# tipagem dos atributos

dados.dtypes

NU_INSCRICAO                object
CO_MUNICIPIO_RESIDENCIA      int64
NO_MUNICIPIO_RESIDENCIA     object
CO_UF_RESIDENCIA             int64
SG_UF_RESIDENCIA            object
NU_IDADE                     int64
TP_SEXO                     object
TP_ESTADO_CIVIL              int64
TP_COR_RACA                  int64
TP_NACIONALIDADE             int64
CO_MUNICIPIO_NASCIMENTO    float64
NO_MUNICIPIO_NASCIMENTO     object
CO_UF_NASCIMENTO           float64
SG_UF_NASCIMENTO            object
TP_ST_CONCLUSAO              int64
TP_ANO_CONCLUIU              int64
TP_ESCOLA                    int64
TP_ENSINO                  float64
IN_TREINEIRO                 int64
CO_ESCOLA                  float64
CO_MUNICIPIO_ESC           float64
NO_MUNICIPIO_ESC            object
CO_UF_ESC                  float64
SG_UF_ESC                   object
TP_DEPENDENCIA_ADM_ESC     float64
TP_LOCALIZACAO_ESC         float64
TP_SIT_FUNC_ESC            float64
TP_PRESENCA_CN               int64
TP_PRESENCA_CH      

In [13]:
# excluir atributos

dados1 = dados.drop(columns=['CO_MUNICIPIO_RESIDENCIA'])

In [14]:
dados1.head()

NU_INSCRICAO NO_MUNICIPIO_RESIDENCIA  CO_UF_RESIDENCIA SG_UF_RESIDENCIA  \
0     1,90E+11            Santa Isabel                35               SP   
1     1,90E+11             Pinhalzinho                35               SP   
2     1,90E+11               São Paulo                35               SP   
3     1,90E+11   São Bernardo do Campo                35               SP   
4     1,90E+11    São Joaquim da Barra                35               SP   

   NU_IDADE TP_SEXO  TP_ESTADO_CIVIL  TP_COR_RACA  TP_NACIONALIDADE  \
0        22       M                1            3                 1   
1        19       M                1            3                 1   
2        17       F                1            3                 1   
3        19       M                1            3                 1   
4        37       M                2            1                 1   

   CO_MUNICIPIO_NASCIMENTO  ...  NU_NOTA_MT  TP_LINGUA TP_STATUS_REDACAO  \
0                1506500.0  ...      6002.0          0               1.0   
1                1503804.0  ...         NaN          1               NaN   
2                1501402.0  ...      7318.0          0               1.0   
3                3548708.0  ...      3932.0          1               1.0   
4                3549409.0  ...         NaN          1               NaN   

   NU_NOTA_COMP1  NU_NOTA_COMP2  NU_NOTA_COMP3  NU_NOTA_COMP4  NU_NOTA_COMP5  \
0          160.0          200.0          180.0          200.0          200.0   
1            NaN            NaN            NaN            NaN            NaN   
2          160.0          120.0          200.0          200.0          200.0   
3           80.0          160.0          120.0          100.0          100.0   
4            NaN            NaN            NaN            NaN            NaN   

   NU_NOTA_REDACAO  Q025  
0            940.0     B  
1              NaN     A  
2            880.0     A  
3            560.0     B  
4              NaN     B  

[5 rows x 43 columns]

In [16]:
# excluir mais de um atributo

dados1 = dados1.drop(columns=['NO_MUNICIPIO_NASCIMENTO','CO_UF_NASCIMENTO','SG_UF_NASCIMENTO',
                              'TP_ANO_CONCLUIU','TP_ENSINO','CO_MUNICIPIO_ESC','CO_UF_ESC','SG_UF_ESC'])

In [17]:
dados1.head()

NU_INSCRICAO NO_MUNICIPIO_RESIDENCIA  CO_UF_RESIDENCIA SG_UF_RESIDENCIA  \
0     1,90E+11            Santa Isabel                35               SP   
1     1,90E+11             Pinhalzinho                35               SP   
2     1,90E+11               São Paulo                35               SP   
3     1,90E+11   São Bernardo do Campo                35               SP   
4     1,90E+11    São Joaquim da Barra                35               SP   

   NU_IDADE TP_SEXO  TP_ESTADO_CIVIL  TP_COR_RACA  TP_NACIONALIDADE  \
0        22       M                1            3                 1   
1        19       M                1            3                 1   
2        17       F                1            3                 1   
3        19       M                1            3                 1   
4        37       M                2            1                 1   

   CO_MUNICIPIO_NASCIMENTO  ...   NU_NOTA_MT  TP_LINGUA  TP_STATUS_REDACAO  \
0                1506500.0  ...       6002.0          0                1.0   
1                1503804.0  ...          NaN          1                NaN   
2                1501402.0  ...       7318.0          0                1.0   
3                3548708.0  ...       3932.0          1                1.0   
4                3549409.0  ...          NaN          1                NaN   

   NU_NOTA_COMP1 NU_NOTA_COMP2  NU_NOTA_COMP3  NU_NOTA_COMP4  NU_NOTA_COMP5  \
0          160.0         200.0          180.0          200.0          200.0   
1            NaN           NaN            NaN            NaN            NaN   
2          160.0         120.0          200.0          200.0          200.0   
3           80.0         160.0          120.0          100.0          100.0   
4            NaN           NaN            NaN            NaN            NaN   

   NU_NOTA_REDACAO  Q025  
0            940.0     B  
1              NaN     A  
2            880.0     A  
3            560.0     B  
4              NaN     B  

[5 rows x 35 columns]

In [18]:
dados1 = dados1.drop(columns=['TP_DEPENDENCIA_ADM_ESC','TP_LOCALIZACAO_ESC','TP_SIT_FUNC_ESC'],)

In [19]:
dados1.head()

NU_INSCRICAO NO_MUNICIPIO_RESIDENCIA  CO_UF_RESIDENCIA SG_UF_RESIDENCIA  \
0     1,90E+11            Santa Isabel                35               SP   
1     1,90E+11             Pinhalzinho                35               SP   
2     1,90E+11               São Paulo                35               SP   
3     1,90E+11   São Bernardo do Campo                35               SP   
4     1,90E+11    São Joaquim da Barra                35               SP   

   NU_IDADE TP_SEXO  TP_ESTADO_CIVIL  TP_COR_RACA  TP_NACIONALIDADE  \
0        22       M                1            3                 1   
1        19       M                1            3                 1   
2        17       F                1            3                 1   
3        19       M                1            3                 1   
4        37       M                2            1                 1   

   CO_MUNICIPIO_NASCIMENTO  ...   NU_NOTA_MT  TP_LINGUA  TP_STATUS_REDACAO  \
0                1506500.0  ...       6002.0          0                1.0   
1                1503804.0  ...          NaN          1                NaN   
2                1501402.0  ...       7318.0          0                1.0   
3                3548708.0  ...       3932.0          1                1.0   
4                3549409.0  ...          NaN          1                NaN   

   NU_NOTA_COMP1 NU_NOTA_COMP2  NU_NOTA_COMP3  NU_NOTA_COMP4  NU_NOTA_COMP5  \
0          160.0         200.0          180.0          200.0          200.0   
1            NaN           NaN            NaN            NaN            NaN   
2          160.0         120.0          200.0          200.0          200.0   
3           80.0         160.0          120.0          100.0          100.0   
4            NaN           NaN            NaN            NaN            NaN   

   NU_NOTA_REDACAO  Q025  
0            940.0     B  
1              NaN     A  
2            880.0     A  
3            560.0     B  
4              NaN     B  

[5 rows x 32 columns]

## 2.1 Corrigindo erros das notas

In [20]:
dados1.loc[:,'NU_NOTA_CN'] /= 10

In [21]:
dados1.head()

NU_INSCRICAO NO_MUNICIPIO_RESIDENCIA  CO_UF_RESIDENCIA SG_UF_RESIDENCIA  \
0     1,90E+11            Santa Isabel                35               SP   
1     1,90E+11             Pinhalzinho                35               SP   
2     1,90E+11               São Paulo                35               SP   
3     1,90E+11   São Bernardo do Campo                35               SP   
4     1,90E+11    São Joaquim da Barra                35               SP   

   NU_IDADE TP_SEXO  TP_ESTADO_CIVIL  TP_COR_RACA  TP_NACIONALIDADE  \
0        22       M                1            3                 1   
1        19       M                1            3                 1   
2        17       F                1            3                 1   
3        19       M                1            3                 1   
4        37       M                2            1                 1   

   CO_MUNICIPIO_NASCIMENTO  ...   NU_NOTA_MT  TP_LINGUA  TP_STATUS_REDACAO  \
0                1506500.0  ...       6002.0          0                1.0   
1                1503804.0  ...          NaN          1                NaN   
2                1501402.0  ...       7318.0          0                1.0   
3                3548708.0  ...       3932.0          1                1.0   
4                3549409.0  ...          NaN          1                NaN   

   NU_NOTA_COMP1 NU_NOTA_COMP2  NU_NOTA_COMP3  NU_NOTA_COMP4  NU_NOTA_COMP5  \
0          160.0         200.0          180.0          200.0          200.0   
1            NaN           NaN            NaN            NaN            NaN   
2          160.0         120.0          200.0          200.0          200.0   
3           80.0         160.0          120.0          100.0          100.0   
4            NaN           NaN            NaN            NaN            NaN   

   NU_NOTA_REDACAO  Q025  
0            940.0     B  
1              NaN     A  
2            880.0     A  
3            560.0     B  
4              NaN     B  

[5 rows x 32 columns]

In [22]:
dados1.loc[:,'NU_NOTA_CH'] /= 10
dados1.loc[:,'NU_NOTA_LC'] /= 10
dados1.loc[:,'NU_NOTA_MT'] /= 10

## 2.2 Renomeando variáveis

In [23]:
dados1 = dados1.rename(columns={'NU_NOTA_REDACAO': 'NOTA_REDACAO'})

In [24]:
dados1 = dados1.rename(columns={'NU_NOTA_CN': 'NOTA_CN','NU_NOTA_CH': 'NOTA_CH',
                                'NU_NOTA_LC': 'NOTA_LC','NU_NOTA_MT': 'NOTA_MT'})

In [25]:
dados1 = dados1.rename(columns={'NU_NOTA_COMP1': 'COMP1','NU_NOTA_COMP2': 'COMP2',
                                'NU_NOTA_COMP3': 'COMP3','NU_NOTA_COMP4': 'COMP4',
                                'NU_NOTA_COMP5':'COMP5'})

In [26]:
dados1 = dados1.rename(columns={'NU_IDADE': 'IDADE','TP_SEXO': 'SEXO',
                                'TP_COR_RACA': 'RACA','Q025': 'INTERNET',
                                'TP_ESCOLA':'ESCOLA'})

## 2.3 Renomeando registros em linha

In [27]:
dados1['RACA'] = dados1['RACA'].replace({0:"nao_declarado", 1:"branca", 2:"preta", 3:"parda", 4:"amarela", 5:"indigena"})

In [29]:
dados1['RACA']

0           parda
1           parda
2           parda
3           parda
4          branca
5           parda
6          branca
7          branca
8          branca
9           parda
10         branca
11        amarela
12          parda
13         branca
14          parda
15          parda
16        amarela
17         branca
18          parda
19          parda
20          parda
21          parda
22          parda
23          parda
24          preta
25         branca
26          parda
27          parda
28          preta
29         branca
           ...   
813742     branca
813743      preta
813744      parda
813745     branca
813746     branca
813747      parda
813748     branca
813749      parda
813750     branca
813751     branca
813752     branca
813753     branca
813754     branca
813755      parda
813756     branca
813757     branca
813758     branca
813759      parda
813760     branca
813761      parda
813762      parda
813763      parda
813764      parda
813765     branca
813766    

In [30]:
dados1['TP_LINGUA'] = dados1['TP_LINGUA'].replace({0:"ingles", 1:"espanhol"})

In [31]:
dados1['ESCOLA'] = dados1['ESCOLA'].replace({1:"nao respondeu", 2:"publica", 3:"privada", 4:"exterior"})

In [32]:
dados1['INTERNET'] = dados1['INTERNET'].replace({"A": "Nao", "B":"Sim"})

# 3. Análise exploratória dos dados

In [38]:
dados1['IDADE'].value_counts().sort_index()

2          1
3          1
4          1
5          2
6          2
7          3
8          6
12         6
13        29
14       355
15     11583
16     59316
17    219313
18    147548
19     81236
20     53009
21     37242
22     29489
23     23531
24     18851
25     15103
26     12465
27     10367
28      8986
29      7650
30      6967
31      6331
32      5685
33      5175
34      4795
       ...  
55       652
56       530
57       497
58       388
59       325
60       254
61       236
62       176
63       144
64       139
65       115
66        81
67        72
68        53
69        43
70        40
71        25
72        17
73        16
74        13
75        10
76         8
77         3
78         3
79         4
80         2
81         3
82         2
86         1
91         1
Name: IDADE, Length: 80, dtype: int64

## 3.1 Análise de valores estranhos: menores de 12 anos

In [39]:
menores_12 = dados1.query('IDADE < 12')['NO_MUNICIPIO_RESIDENCIA'].value_counts()

In [40]:
menores_12.head()

São Paulo                 5
Vargem Grande Paulista    1
Santana de Parnaíba       1
Araçatuba                 1
Ribeirão Preto            1
Name: NO_MUNICIPIO_RESIDENCIA, dtype: int64

In [41]:
dados_maiores_11 = dados1.loc[dados1['IDADE']>11]

In [42]:
dados_maiores_11.head()

NU_INSCRICAO NO_MUNICIPIO_RESIDENCIA  CO_UF_RESIDENCIA SG_UF_RESIDENCIA  \
0     1,90E+11            Santa Isabel                35               SP   
1     1,90E+11             Pinhalzinho                35               SP   
2     1,90E+11               São Paulo                35               SP   
3     1,90E+11   São Bernardo do Campo                35               SP   
4     1,90E+11    São Joaquim da Barra                35               SP   

   IDADE SEXO  TP_ESTADO_CIVIL    RACA  TP_NACIONALIDADE  \
0     22    M                1   parda                 1   
1     19    M                1   parda                 1   
2     17    F                1   parda                 1   
3     19    M                1   parda                 1   
4     37    M                2  branca                 1   

   CO_MUNICIPIO_NASCIMENTO    ...     NOTA_MT TP_LINGUA  TP_STATUS_REDACAO  \
0                1506500.0    ...       600.2    ingles                1.0   
1                1503804.0    ...         NaN  espanhol                NaN   
2                1501402.0    ...       731.8    ingles                1.0   
3                3548708.0    ...       393.2  espanhol                1.0   
4                3549409.0    ...         NaN  espanhol                NaN   

   COMP1  COMP2  COMP3  COMP4  COMP5  NOTA_REDACAO  INTERNET  
0  160.0  200.0  180.0  200.0  200.0         940.0       Sim  
1    NaN    NaN    NaN    NaN    NaN           NaN       Nao  
2  160.0  120.0  200.0  200.0  200.0         880.0       Nao  
3   80.0  160.0  120.0  100.0  100.0         560.0       Sim  
4    NaN    NaN    NaN    NaN    NaN           NaN       Sim  

[5 rows x 32 columns]

In [43]:
dados_maiores_11['IDADE'].value_counts().sort_index()

12         6
13        29
14       355
15     11583
16     59316
17    219313
18    147548
19     81236
20     53009
21     37242
22     29489
23     23531
24     18851
25     15103
26     12465
27     10367
28      8986
29      7650
30      6967
31      6331
32      5685
33      5175
34      4795
35      4340
36      4109
37      4105
38      3767
39      3299
40      3058
41      2638
       ...  
55       652
56       530
57       497
58       388
59       325
60       254
61       236
62       176
63       144
64       139
65       115
66        81
67        72
68        53
69        43
70        40
71        25
72        17
73        16
74        13
75        10
76         8
77         3
78         3
79         4
80         2
81         3
82         2
86         1
91         1
Name: IDADE, Length: 73, dtype: int64

## 3.2 Treineiros

In [47]:
dados_maiores_11.query('IN_TREINEIRO==1')['IN_TREINEIRO'].value_counts()

1    93988
Name: IN_TREINEIRO, dtype: int64

In [48]:
treineiros = dados_maiores_11.loc[dados_maiores_11['IN_TREINEIRO']==1]

In [49]:
treineiros.to_csv('treineiros_enem_2019_sp.csv', encoding='iso-8859-1', index=False)

## 3.3 Vestibulandos

In [50]:
vestibulandos = dados_maiores_11.loc[dados_maiores_11['IN_TREINEIRO']==0]

In [52]:
vestibulandos.to_csv('vestibulandos_enem_2019_sp.csv', encoding='iso-8859-1', index=False)

## 3.4 Análise das presenças

In [53]:
vestibulandos.groupby('TP_PRESENCA_CN')['TP_PRESENCA_CN'].count()

TP_PRESENCA_CN
0    218693
1    500805
2       270
Name: TP_PRESENCA_CN, dtype: int64

In [54]:
vestibulandos.groupby('TP_PRESENCA_CH')['TP_PRESENCA_CH'].count()

TP_PRESENCA_CH
0    185969
1    533142
2       657
Name: TP_PRESENCA_CH, dtype: int64

In [56]:
vestibulandos.groupby('TP_PRESENCA_LC')['TP_PRESENCA_LC'].count()

TP_PRESENCA_LC
0    185969
1    533142
2       657
Name: TP_PRESENCA_LC, dtype: int64

In [57]:
vestibulandos.groupby('TP_PRESENCA_MT')['TP_PRESENCA_MT'].count()

TP_PRESENCA_MT
0    218693
1    500805
2       270
Name: TP_PRESENCA_MT, dtype: int64

In [59]:
vestibulandos.groupby('TP_STATUS_REDACAO')['TP_STATUS_REDACAO'].count()

TP_STATUS_REDACAO
1.0    519763
2.0       913
3.0      1417
4.0      5513
6.0      3626
7.0       543
8.0       828
9.0       539
Name: TP_STATUS_REDACAO, dtype: int64

## 3.5 Valores ausentes

In [60]:
vestibulandos['IDADE'].isnull().sum()

0

In [61]:
provas = ['NOTA_REDACAO', 'NOTA_CN','NOTA_MT','NOTA_LC','NOTA_CH']

In [62]:
vestibulandos[provas].isnull().sum()

NOTA_REDACAO    186626
NOTA_CN         218963
NOTA_MT         218963
NOTA_LC         186626
NOTA_CH         186626
dtype: int64

In [63]:
presenca_total = ['TP_PRESENCA_CN','TP_PRESENCA_CH','TP_PRESENCA_MT','TP_PRESENCA_LC']

In [64]:
vestibulandos['presenca_total'] = vestibulandos[presenca_total].sum(axis=1)

C:\Program Files (x86)\Microsoft Visual Studio\Shared\Anaconda3_64\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [65]:
vestibulandos.head()

NU_INSCRICAO NO_MUNICIPIO_RESIDENCIA  CO_UF_RESIDENCIA SG_UF_RESIDENCIA  \
0     1,90E+11            Santa Isabel                35               SP   
1     1,90E+11             Pinhalzinho                35               SP   
2     1,90E+11               São Paulo                35               SP   
3     1,90E+11   São Bernardo do Campo                35               SP   
4     1,90E+11    São Joaquim da Barra                35               SP   

   IDADE SEXO  TP_ESTADO_CIVIL    RACA  TP_NACIONALIDADE  \
0     22    M                1   parda                 1   
1     19    M                1   parda                 1   
2     17    F                1   parda                 1   
3     19    M                1   parda                 1   
4     37    M                2  branca                 1   

   CO_MUNICIPIO_NASCIMENTO       ...        TP_LINGUA TP_STATUS_REDACAO  \
0                1506500.0       ...           ingles               1.0   
1                1503804.0       ...         espanhol               NaN   
2                1501402.0       ...           ingles               1.0   
3                3548708.0       ...         espanhol               1.0   
4                3549409.0       ...         espanhol               NaN   

   COMP1  COMP2  COMP3  COMP4  COMP5  NOTA_REDACAO  INTERNET  presenca_total  
0  160.0  200.0  180.0  200.0  200.0         940.0       Sim               4  
1    NaN    NaN    NaN    NaN    NaN           NaN       Nao               0  
2  160.0  120.0  200.0  200.0  200.0         880.0       Nao               4  
3   80.0  160.0  120.0  100.0  100.0         560.0       Sim               4  
4    NaN    NaN    NaN    NaN    NaN           NaN       Sim               0  

[5 rows x 33 columns]

- Se a soma for 0 faltou nos dois dias
- Se a soma for 2, foi em um dia
- Se a soma for 4, foi nos dois dias
- Se a soma for 6 ou 8 teve desclassificação.

In [66]:
vestibulandos.groupby('presenca_total')['presenca_total'].count()

presenca_total
0    185175
2     33726
4    500526
6       341
Name: presenca_total, dtype: int64

COMPARECIMENTO NOS DIAS DAS PROVAS:

- 500526 inscritos compareceram nos dois dias.

- 185175 inscritos faltaram nos dois dias.

- 33726 inscritos faltaram em um dos dias.

- 341 foram desclassificados.

Utilizando apenas os alunos presentes nos dois dias

In [67]:
vestibulandos_presentes = vestibulandos.loc[vestibulandos['presenca_total']==4]

Filtrando apenas os vestibulandos sem irregularidades

In [68]:
vestibulandos_classificados = vestibulandos_presentes.loc[vestibulandos_presentes['TP_PRESENCA_CN']==1]

In [71]:


vestibulandos_classificados[['NOTA_REDACAO','NOTA_CN','NOTA_MT','NOTA_LC','NOTA_CH']].isnull().sum()

NOTA_REDACAO    0
NOTA_CN         0
NOTA_MT         0
NOTA_LC         0
NOTA_CH         0
dtype: int64

## 3.6 Notas zero das redações

In [82]:
vestibulandos_classificados[vestibulandos_classificados['NOTA_REDACAO']==0].groupby('NOTA_REDACAO')['NOTA_REDACAO'].count()

NOTA_REDACAO
0.0    9181
Name: NOTA_REDACAO, dtype: int64